### ISSUE TITLE
Support Slots

### ISSUE THREAD

**Issue by msaraiva** [phoenixframework/phoenix_live_view#1617]
Initial proposal to bring slots to LV components.

## Motivation

Slots provide a standard way to pass blocks of templates to components, improving composability. The blocks can also hold additional  information (slot props) that can be used by the component to customize the rendering.

The concept is similar to component properties as it defines a way to pass information to a component using its public API. The main difference is that slots can only be assigned as direct children of the component instead of attributes.

### Example (Surface)

The component definition:

```elixir
defmodule Card do
  use Surface.Component

  slot header
  slot footer
  slot default, required: true

  def render(assigns) do
    ~F"""
    <div class="card">
      <header class="card-header">
        <p class="card-header-title">
          <#slot name="header" />
        </p>
      </header>
      <div class="card-content">
        <div class="content">
          <#slot />
        </div>
      </div>
      <footer class="card-footer">
        <#slot name="footer" />
      </footer>
    </div>
    """
  end
end
```

Usage:

```jsx
<Card>
  <:header>
    A simple card component
  </:header>

  This example demonstrates how to create components with multiple slots.
  It defines a <strong>default</strong> slot to hold the card's content
  and two <strong>named slots</strong>: header and footer.

  <:footer>
    <a href="#" class="card-footer-item">Footer Item 1</a>
    <a href="#" class="card-footer-item">Footer Item 2</a>
  </:footer>
</Card>
```

## Current implementation

Since there's no built-in support for slots in LV, Surface's implementation depends on pattern matching the arguments passed by `render_block/2`. Something like the following (in HEEx):

```elixir
<%= component &Card.render/1 do %>
  <% {:default, 0, _args} -> %>
    This example...

  <% {:header, 0, _args} -> %>
    A simple card component

  <% {:footer, 0, _args} -> %>
    <a href="#" class="card-footer-item">Footer Item 1</a>
    ...
<% end %>
```

The assigned slot item is represented as `{name, index, args}` where `name` is the name of the slot, `index` is the index of the assigned block (as it accepts multiple blocks for the same slot) and `args` are the values passed as the second argument of `render_block/2`.

## Limitation

One limitation of this approach it that, in order to make the change tracker work nicely, the clauses representing each slot item must be generated at compile-time, that means the user can't assign slots dynamically, for instance using a `for` comprehension.

By having slots as a built-in feature with a standard API we can improve composability and provide better ergonomics to HEEx. It would also not on only  keep change tracking working but it would allow further optimizations as the engine will be aware of the slots API.

## Solution

The standard API could implement slots as regular assigns (probably a list of maps) where each map in that list would represent a single assigned slot item. Each item should hold the passed values of the slot props along with a `inner_block` field containing the anonymous function representing the inner block of that slot item. Each inner block could be rendered using an updated version of `render_block` or something similar.

On top of that, we should extend the HTML tokenizer/engine to accept a custom syntax to assign slots. The syntax could be the same as the Surface's syntax or whatever the core team think is better for HEEx. As long as the underlying model covers all features, the syntax is just a matter of taste. 

## More info

* https://surface-ui.org/slots
* https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_templates_and_slots
* https://v3.vuejs.org/guide/component-slots.html

**Comment by josevalim** [comment]
@msaraiva I am :+1: on the proposal as a whole! I like Surface's syntax for passing slots with `<:slot></:slot>`, because they are assigns after all and therefore atoms.

I would skip the `<#slot name="header" />` syntax on the initial version. I think we most likely want to have it, but I want to focus on the low-level API first. The low-level API is important because that's what we will optimize and what will be reused by Surface, so I want to make sure we nail it before we think on the high-level sugar.

> The standard API could implement slots as regular assigns (probably a list of maps) where each map in that list would represent a single assigned slot item.

What do you think here?

    [%{name: atom, assigns: map, inner_block: fun.() | nil}]

**Comment by msaraiva** [comment]
> I would skip the `<#slot name="header" />` syntax on the initial version

I agree.

> What are you thinking here? Something like this?
> 
> ```
> [%{name: atom, assigns: map, inner_block: fun.() | nil}]
> ```

Yeah. Something like that. Not sure if using a tuple instead of a map would be better.

**Comment by josevalim** [comment]
Definitely a map, it will allow us to add more fields in the future if necessary. :) I think you can go ahead with the implementation and once we have an initial version I can jump into the diff tracking engine!

**Comment by msaraiva** [comment]
Closed in #1661.

**Comment by formido** [comment]
@msaraiva Am I right in thinking you can't dynamically add slots yet? I get "A slot entry must be a direct child of a component".
